api/nodes/permissions.py

```diff
@@ -15,6 +17,34 @@ def has_object_permission(self, request, view, obj):
             return obj.can_edit(auth)
 
 
+class AdminOrPublic(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be a Node or User, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
+        if request.method in permissions.SAFE_METHODS:
+            return node.is_public or node.can_view(auth)
+        else:
+            return node.has_permission(auth.user, osf_permissions.ADMIN)
+
+
+class ContributorDetailPermissions(permissions.BasePermission):
+    """Permissions for contributor detail page."""
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, User)), 'obj must be User or Node, got {}'.format(obj)
+        auth = get_user_auth(request)
+        node = Node.load(request.parser_context['kwargs']['node_id'])
```

Comment:
In any case, it wouldn't hurt to document the contract in the docstring, e.g "This can only be applied to view classes that define node_lookup_url_kwarg and user_lookup_url_kwarg".

Not absolutely necessary, but it's nice.